### PR TITLE
Implement tagged and dynamic metrics

### PR DESCRIPTION
**Pull request by awjchen**
This PR adds support for tagged and dynamic metrics, via a strongly-typed interface based on the [monad-metrics-extensible](https://hackage.haskell.org/package/monad-metrics-extensible) package.

Haddocks for the new interface may be temporarily found [here](https://hasura.github.io/temp-new-ekg-api/ekg-core/System-Metrics.html). A tutorial for this new interface has also been written -- see [Tutorial.md](https://github.com/hasura/ekg-core/blob/alex/newApi-5/Tutorial.md).

This PR also addresses issues 24, 29, 31, and 41 of tibbe/ekg-core, incorporating work from PRs 25, 30, and 42 of tibbe/ekg-core.

This PR also drops support for older versions of GHC using `base < 4.10`.

Related PRs:
- https://github.com/hasura/ekg-json/pull/2
- https://github.com/hasura/ekg/pull/1
- https://github.com/hasura/ekg-statsd/pull/2

### CODE REVIEW

File: System/Metrics.hs
Comment (paf31) [inline System/Metrics.hs:322]: Might be worth adding an example in the published comments along with a justification (in a non-published comment) for why `coerce` is a safe implementation.

Comment (awjchen) [reply System/Metrics.hs:322]: Sounds good!

Though, I suppose this function is not safe because it never evaluates its first argument. This allows a user to change scope arbitrarily via `subset undefined`. I don't know how to prevent the use of `undefined` ... I was hoping that we could trust users to not use `undefined`, or maybe allow them to use it as an escape hatch.

Comment (paf31) [reply System/Metrics.hs:322]: Thanks for adding the example - can you please explain the intended purpose of the scopes/subset functionality though, because I'm not seeing it. Are we using it?

File: Data/Atomic.hs
Comment (paf31) [inline Data/Atomic.hs:22]: I'm curious about the need for optimizing this so carefully here. It's worth adding a comment at least about why we need to optimize this so much. My intuition is that for our needs, this could be premature optimization, given that the time to compute some estimators is probably a drop in the ocean, given the amount of JSON serialization and network IO that we're doing in the course of reporting this stuff. I could believe it's needed though, so any elucidation in comments here would be appreciated.

File: System/Metrics/Distribution.hs
Comment (jkachmar) [inline System/Metrics/Distribution.hs:140]: I'm always a little wary whenever I see spinlocks, but I'm not 100% sure about the constraints you're working with here so in lieu of offering targeted advice I'll defer to [this blog post by Aleksey Kladov](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html) as well as his [follow-up indicating that "sleeping mutexes" should be faster anyway](https://matklad.github.io/2020/01/04/mutexes-are-faster-than-spinlocks.html).

Comment (awjchen) [reply System/Metrics/Distribution.hs:140]: Thanks for showing me the dangers of spinlocks! To be honest, I don't know what I'm doing when it comes to spinlocks or locking in general. This code is a translation of and replacement for some C code that was previously in this `ekg-core` library (and the replacement was motivated by reports that the C code was generating nonsensical results on aarch64). But now I think I might need to find an alternative locking mechanism ...

File: Data/Atomic.hs
Comment (awjchen) [reply Data/Atomic.hs:22]: Hmmmm. I don't have much knowledge or experience about optimization, so I might not be very helpful here. @jberryman, could I have your take on this?

The use of fetch-and-add here is inherited from the replaced `ekg-core` code, the implementation of which was motivated by reports of contention in `ekg` under an old implementation that used `atomicModifyIORef'` -- see the last section of [Tibell's post](https://blog.johantibell.com/2014/05/ekg-04-released.html). Indeed, repeating Tibell's contention-heavy micro-benchmark, fetch-and-add was over 100x faster than `atomicModifyIORef` and about 5x as fast as [`atomicModifyIORefCas_`](https://hackage.haskell.org/package/atomic-primops-0.8.4/docs/Data-Atomics.html#v:atomicModifyIORefCAS_). It seems highly unlikely that our use case would trigger that level of contention, but there seems also seems to be little downside to using fetch-and-add here.

Padding the counter to 64 bytes is new -- I copied that technique from `unagi-chan`. However, unlike fetch-and-add, it has the downside that it uses additional memory, so this might be premature optimization. @jberryman was also skeptical about this padding when I showed to him a while back. It would certainly do no harm to our use case if we removed this padding.

In any case, I'll add a comment.

File: System/Metrics.hs
Comment (awjchen) [reply System/Metrics.hs:322]: Sure! I had two motivations when introducing these "scopes"/"subsets".

(a) I wanted to preserve an `ekg-core` feature in the presence of the statically-defined metrics introduced in this fork. The `ekg-core` feature is that users can register to their store metrics defined by other code (i.e. libraries). Since this fork requires that all metrics be statically defined, the library code must register their metrics under some metrics specification. "Subsets" were a convenient way for users to include those library-defined metrics specifications in their own metrics specifications, thereby allowing users to register library-defined metrics. I didn't think further than this, though. Perhaps there is another solution not involving these "scopes".

In my integration of this fork with our codebase, I use "scopes" to register the pre-defined RTS metrics provided by `ekg-core`, and to register the `ServerMetrics` defined by `graphql-engine`.

(b) `ekg-core` supports static metrics only, but this fork also supports dynamic metrics (that is, metrics can be deregistered and re-registered). This support for dynamic metrics expands the number of programs that compile and run without error. For example, in `ekg-core`, registering a metric that already exists results in an exception, whereas in this fork the existing metric would be replaced, since that is now a valid operation.

Using "scopes" was a way to recover the safety of static metrics for those metrics that don't need to be -- and shouldn't be -- deregistered. To make static metrics with "scopes", you would 1. separate your dynamic and static metrics into two "subsets" of your metrics specification, 2. register the static metrics at startup, then 3. create a new reference to the store scoped only to the dynamic metrics and discard the original store reference. This statically guarantees that the metrics in the "static" subset can no longer be deregistered.

In my integration of this fork with our codebase, I used this "technique", but it is not essential.

Does all this sound reasonable to you?

File: Tutorial.md
Comment (evertedsphere) [inline Tutorial.md]: s/trt/tr/

File: Tutorial.md
Comment (evertedsphere) [inline Tutorial.md]: s/rei/regi/

File: Tutorial.md
Comment (evertedsphere) [inline Tutorial.md]: it'd help to rewrite the first clause to explain that sampling multiple metrics individually in sequence is not guaranteed to be atomic — atomicity or a lack thereof isn't defined in the context of sampling a single primitive value, right?

File: Tutorial.md
Comment (evertedsphere) [inline Tutorial.md]: perhaps
```diff
- metrics may be annotated by sets of key-value pairs called **tags** or **tag sets**.
+ metrics may be annotated by a **tag set**, which is a set 
+ of key-value pairs called **tags**.
```

**Review by evertedsphere — APPROVED** [review]
looks good to me! 

love the direction you chose when designing the new api. outside of that, the store action tests are also really cool, and the tutorial is quite well-written too

(left some nitpicks over typos etc, feel free to delete them or mark them resolved once you've gone over them)

File: Tutorial.md
Comment (awjchen) [reply Tutorial.md]: This is great!

Comment (evertedsphere) [reply Tutorial.md]: :)

File: Tutorial.md
Comment (awjchen) [reply Tutorial.md]: :pray:

File: Tutorial.md
Comment (awjchen) [reply Tutorial.md]: :pray:

File: Tutorial.md
Comment (awjchen) [reply Tutorial.md]: This is true, though ... I think I completely botched this sentence :sweat_smile: It doesn't really communicate the point I wanted to make, so I've rewritten it now.

Comment (evertedsphere) [reply Tutorial.md]: ah, sampleAll is not atomic, that makes sense (because it's multiple metrics)

Comment (evertedsphere) [reply Tutorial.md]: why not make `sampleAll` atomic by default (if that's possible)?

Comment (evertedsphere) [reply Tutorial.md]: or, well, rename it to `sampleAllNonAtomic` or something, and recommend using an atomic `sampleAll` function by default unless the user is happy to be provided a possibly-internally-inconsistent snapshot of the metrics in which they can use the non-atomic function instead

Comment (awjchen) [reply Tutorial.md]: Hmmmm, I think that, if the sampling actions we accept from the users can be arbitrary `IO` actions, it's not possible for us to guarantee the atomicity of a sample containing multiple metrics.

Even when we have an atomically-sampled group of metrics registered via `registerGroup`, we are relying on the user to give us an `IO` action that atomically produces the group's shared value. In that sense, the `registerGroup` function just allows a user to express a value that _they_ compute atomically in a format that `ekg-core` can understand (i.e. by projecting that value onto multiple metrics).

Comment (evertedsphere) [reply Tutorial.md]: ah i get it now. i think the docs could be improved to explain this better, just the way you did above, perhaps?

Comment (awjchen) [reply Tutorial.md]: I've updated the tutorial to include the bit about why sampling isn't atomic!

File: System/Metrics/Distribution.hs
Comment (awjchen) [reply System/Metrics/Distribution.hs:140]: I would like to postpone replacing the spinlock. Rather than doing any more work on the `Distribution` metric type, I think we should replace it entirely, maybe with an implementation of Prometheus' histograms and summaries?

File: System/Metrics.hs
Comment (awjchen) [reply System/Metrics.hs:322]: I've added source code comments explaining why `subset`'s implementation is safe, and added a section to the tutorial showing how to simulate static metrics (what I tried to explain in the above part (b)).

Although the scopes/subsets may be a sub-optimal API and may need to be revised or replaced, we will have the opportunity to do so because this library will only be used internally for some time. Then, I'm going to go ahead and merge the current API with scopes/subsets with the intention of returning to it later.

File: Data/Atomic.hs
Comment (awjchen) [reply Data/Atomic.hs:22]: I've added commits removing the padding for the counter metric and justifying it for the distribution metric.

**Comment by awjchen** [comment]
This PR now also has approval from @jberryman. Merging now. Also merging the other PRs.

**Pull request merged**
